src/server/ua_services_discovery_multicast.c

```diff
@@ -101,83 +108,269 @@ addMdnsRecordForNetworkLayer(UA_Server *server, const UA_String *appName,
+#if UA_IPV6
+static UA_StatusCode
+getInterfaceIndex(UA_UInt32 *index, struct sockaddr_storage *interfaceAddr) {
+    UA_StatusCode ret = UA_STATUSCODE_BADINTERNALERROR;
+    struct ifaddrs *ifaceAddr;
+    const UA_UInt32 defaultIfaceIndex = 0;
+
+    struct sockaddr_in6 *in6Addr = (struct sockaddr_in6 *)interfaceAddr;
+    if(memcmp(&in6Addr->sin6_addr, &in6addr_any, sizeof(in6Addr->sin6_addr)) == 0) {
+        *index = defaultIfaceIndex;
+        return UA_STATUSCODE_GOOD;
+    }
 
-#ifdef SO_REUSEPORT
-    UA_setsockopt(s, SOL_SOCKET, SO_REUSEPORT, (char *)&flag, sizeof(flag));
+    if(getifaddrs(&ifaceAddr) < 0) {
+        return ret;
+    }
+
+    for(struct ifaddrs *ifa = ifaceAddr; ifa != NULL; ifa = ifa->ifa_next) {
+        if(ifa->ifa_addr == NULL)
+            continue;
+        if(ifa->ifa_addr->sa_family == interfaceAddr->ss_family) {
+            if(memcmp(&in6Addr->sin6_addr,
+                      &((struct sockaddr_in6 *)ifa->ifa_addr)->sin6_addr,
+                      sizeof(in6Addr->sin6_addr)) == 0) {
+                *index = UA_if_nametoindex(ifa->ifa_name);
+                ret = UA_STATUSCODE_GOOD;
+                break;
+            }
+        }
+    }
+
+    freeifaddrs(ifaceAddr);
+    return ret;
+}
+
+/*
+ * Set the custom outbound IPV6 multicast interface.
+ */
+static UA_StatusCode
+setMulticastInterfaceIpv6(UA_Server *server, UA_SOCKET socket,
+                          struct sockaddr_storage *mdnsSockAddr) {
+    UA_StatusCode retval = UA_STATUSCODE_BADINTERNALERROR;
+    UA_UInt32 interface = 0;
+
+    if(server->config.mdnsInterfaceIP.length > 0) {
+        retval = getInterfaceIndex(&interface, mdnsSockAddr);
+        if(retval != UA_STATUSCODE_GOOD)
+            return retval;
+    }
+
+    if(UA_setsockopt(socket, IPPROTO_IPV6, IPV6_MULTICAST_IF, &interface,
+                     sizeof(interface)) < 0) {
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                     "Multicast DNS: failed setting IPV6_MULTICAST_IF %s",
+                     strerror(errno));
+        return retval;
+    }
+    return UA_STATUSCODE_GOOD;
+}
 #endif
-    UA_setsockopt(s, SOL_SOCKET, SO_REUSEADDR, (char *)&flag, sizeof(flag));
-    if(UA_bind(s, (struct sockaddr *)&in, sizeof(in))) {
-        UA_close(s);
+
+static UA_StatusCode
+setMulticastInterfaceIpv4(UA_Server *server, UA_SOCKET socket,
+                          struct sockaddr_storage *mdnsSockAddr) {
+    if(UA_setsockopt(socket, IPPROTO_IP, IP_MULTICAST_IF,
+                     &((struct sockaddr_in *)mdnsSockAddr)->sin_addr,
+                     sizeof(struct in_addr)) < 0) {
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                     "Multicast DNS: failed setting IP_MULTICAST_IF %s", strerror(errno));
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+setMulticastInterface(UA_Server *server, UA_SOCKET sock,
```

Comment:
The near-identical logic was implemented for the UDP EventSource for the EventLoop.
You could at runtime
- get the EventSource by its „udp“ protocol identifier
- Open a connection to the UDP multicast group. This is detected automatically from the IP range.
- Cast the generated connection-identifier to a socket when needed for the mdns library (only for Linux, Win32 sockets are pointers).

I really want to avoid duplicate logic for the same functionality.
If you cannot follow the above steps, please document why. So we can fix the underlying issue in the future.